### PR TITLE
add hookui-lib + changes for support

### PR DESCRIPTION
**Pull request by Captain-Of-Coit**
- Enables easy communication between UIs and C# mods, via built-in `Binding`, `UISystemBase` and `window.engine.on/.trigger`
- Makes it possible to embed JS UI into C# mod assemblies, easier install for players

Missing:

- [x] More testing
- [ ] More docs

Full flow for mod authors would be something like this:

**Define UI**

```javascript
import React from 'react';

const $MyCoolMod = ({ react }) => {
    const [seconds, setSeconds] = react.useState(0);

    react.useEffect(() => {
        const event = "myowncoolmod_namespace.seconds_passed"
        const updateEvent = event + ".update"
        const subscribeEvent = event + ".subscribe"
        const unsubscribeEvent = event + ".unsubscribe"

        var sub = engine.on(updateEvent, (data) => {
            setSeconds(data)
        })
        engine.trigger(subscribeEvent)
        return () => {
            engine.trigger(unsubscribeEvent)
            sub.clear();
        };
    }, [])

    return <div>
                It's been {seconds} seconds since the mod first loaded!
        </div>
}

window._$hookui.registerPanel({
    id: "example.mycoolmod",
    name: "My Cool Mod",
    icon: "Media/Game/Icons/BuildingLevel.svg",
    component: $MyCoolMod
})
```

**Embed resource**

```csharp
<EmbeddedResource Include="dist/mycoolmod.transpiled.js" />
```

**Register Mod UI Extension**

```csharp
    public class MyCoolModUI : UIExtension {
        public new readonly string extensionID = "examples.mycoolmod";
        public new readonly string extensionContent;
        public new readonly ExtensionType extensionType = ExtensionType.Panel;

        public MyCoolModUI() {
            this.extensionContent = this.LoadEmbeddedResource("MyCoolMod.dist.mycoolmod.transpiled.js");
        }
    }
```

**Define UI System, inherit from `UISystemBase`**

```csharp
    class MyUISystem : UISystemBase {
        private int seconds_passed = 0;
        private string kGroup = "myowncoolmod_namespace";
        protected override void OnCreate() {
            base.OnCreate();

            this.AddUpdateBinding(new GetterValueBinding<int>(this.kGroup, "seconds_passed", () => {
                return this.seconds_passed;
            }));

            IncrementPeriodically();
        }
        private async void IncrementPeriodically() {
            while (true) {
                //UnityEngine.Debug.Log($"Adding to {seconds_passed}!");
                await Task.Delay(1000);
                seconds_passed += 1;
            }
        }
    }
```

**Inject system**

```csharp
    [HarmonyPatch(typeof(SystemOrder))]
    public static class SystemOrderPatch {
        [HarmonyPatch("Initialize")]
        [HarmonyPostfix]
        public static void Postfix(UpdateSystem updateSystem) {
            updateSystem.UpdateAt<MyUISystem>(SystemUpdatePhase.UIUpdate);
        }
    }
```

### CODE REVIEW

**Comment by 89pleasure** [comment]
Is this while loop a requirement for the ui to work? Or is it just needed for this example?

**Comment by Captain-Of-Coit** [comment]
> Is this while loop a requirement for the ui to work? Or is it just needed for this example?

No, it's just an example in order to mutate `this.seconds_passed` and demonstrate that the values automatically gets sent to the Game UI when changed.

File: mod/Plugin.cs [hunk 1]
```diff
@@ -214,7 +246,7 @@ public static void InsertScriptTags(string filePath, string destinationPath, Lis
                 scriptTagsBuilder.AppendLine($"<script src=\"{scriptPath}\"></script>");
             }
 
-            fileContent = fileContent.Replace("<!--EXTENSIONS_LIST-->", scriptTagsBuilder.ToString());
+            //fileContent = fileContent.Replace("<!--EXTENSIONS_LIST-->", scriptTagsBuilder.ToString());
```
Comment (Captain-Of-Coit) [inline mod/Plugin.cs:249]: This method needs to be renamed

**Comment by 89pleasure** [comment]
I like it. So I would love to use this soon :)

**Pull request merged**